src/audio/src/src.c

```diff
@@ -891,6 +905,152 @@ static int src_prepare_general(struct comp_dev *dev, struct comp_data *cd)
+		return -EINVAL;
+	}
+
+	cd = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, sizeof(*cd));
+	if (!cd) {
+		rfree(dev);
+		return -ENOMEM;
+	}
+
+	memcpy_s(&cd->ipc_config, sizeof(cd->ipc_config), cfg->init_data, sizeof(cd->ipc_config));
+
+	cd->delay_lines = NULL;
+	cd->src_func = src_fallback;
+	cd->polyphase_func = NULL;
+	src_polyphase_reset(&cd->src);
+
+	md->private = cd;
+
+	mod->simple_copy = true;
+
+	return 0;
+}
+
+static int src_prepare(struct processing_module *mod)
+{
+	struct comp_data *cd = module_get_private_data(mod);
+	struct sof_ipc_stream_params *params = mod->stream_params;
+	struct comp_dev *dev = mod->dev;
+	int ret;
+
+	comp_info(dev, "src_prepare()");
+
+	ret = src_params_general(dev, cd, params);
+	if (ret < 0)
+		return ret;
+
+	return src_prepare_general(dev, cd);
+}
+
+static int src_process(struct processing_module *mod,
+		       struct input_stream_buffer *input_buffers, int num_input_buffers,
+		       struct output_stream_buffer *output_buffers, int num_output_buffers)
+{
+	struct comp_data *cd = module_get_private_data(mod);
+	struct comp_dev *dev = mod->dev;
+	struct comp_buffer *source, *sink;
+	struct comp_buffer __sparse_cache *source_c, *sink_c;
+	int consumed = 0;
+	int produced = 0;
+	int ret;
+
+	comp_dbg(dev, "src_process()");
+
+	/* src component needs 1 source and 1 sink buffer */
+	source = list_first_item(&dev->bsource_list, struct comp_buffer,
+				 sink_list);
+	sink = list_first_item(&dev->bsink_list, struct comp_buffer,
+			       source_list);
+
+	source_c = buffer_acquire(source);
+	sink_c = buffer_acquire(sink);
+
+	ret = src_get_copy_limits(cd, source_c, sink_c);
+	if (ret) {
+		comp_dbg(dev, "No data to process.");
+		goto out;
+	}
+
+	cd->src_func(dev, cd, &source_c->stream, &sink_c->stream, &consumed, &produced);
+	src_update_buffer_position(input_buffers, output_buffers, &consumed, &produced);
+
+	comp_dbg(dev, "src_process(), consumed = %u,  produced = %u", consumed, produced);
+
+out:
+	buffer_release(sink_c);
+	buffer_release(source_c);
+
+	return 0;
+}
+
+static int src_set_config(struct processing_module *mod, uint32_t config_id,
+			  enum module_cfg_fragment_position pos, uint32_t data_offset_size,
+			  const uint8_t *fragment, size_t fragment_size, uint8_t *response,
+			  size_t response_size)
+{
+	return -EINVAL;
+}
+
+static int src_get_config(struct processing_module *mod, uint32_t config_id,
+			  uint32_t *data_offset_size, uint8_t *fragment, size_t fragment_size)
+{
+	return -EINVAL;
+}
+
+static int src_reset(struct processing_module *mod)
+{
+	struct comp_data *cd = module_get_private_data(mod);
+
+	comp_info(mod->dev, "src_reset()");
+
+	cd->src_func = src_fallback;
+	src_polyphase_reset(&cd->src);
+
+	return 0;
+}
+
+static int src_free(struct processing_module *mod)
+{
+	struct comp_data *cd = module_get_private_data(mod);
+
+	comp_info(mod->dev, "src_free()");
+
+	/* Free dynamically reserved buffers for SRC algorithm */
+	if (cd->delay_lines)
```

Comment:
`if` is not needed, `rfree(NULL)` is allowed

Reply:
Thanks, I also added a new commit fixing the previous code.